src/macaron/vsa/vsa.py

```diff
@@ -0,0 +1,251 @@
+# Copyright (c) 2024 - 2024, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""VSA schema and generation."""
+
+from __future__ import annotations
+
+import base64
+import datetime
+import json
+from enum import StrEnum
+from importlib import metadata as importlib_metadata
+from typing import Any, TypedDict
+
+# Note: The lint error "N815:mixedCase variable in class scope" is disabled for
+# field names in the VSA to conform with in-toto naming conventions.
+
+
+class Vsa(TypedDict):
+    """The Macaron Verification Summary Attestation.
+
+    For reference, see:
+    * `SLSA <https://slsa.dev/spec/v1.0/verification_summary>`_.
+    * `in-toto <https://github.com/in-toto/attestation/blob/main/spec/predicates/vsa.md>`_.
+    """
+
+    #: The payload type. Following in-toto, this is always ``application/vnd.in-toto+json``.
+    payloadType: str  # noqa: N815
+
+    #: The payload of the VSA, base64 encoded.
+    payload: str
+
+
+class VsaStatement(TypedDict):
+    """The Statement layer of a Macaron VSA.
+
+    For reference, see:
+    * in-toto Statement layer specification: https://github.com/in-toto/attestation/blob/main/spec/v1/statement.md.
+    """
+
+    _type: str
+    subject: list[dict]
+    predicateType: str  # noqa: N815
+    predicate: VsaPredicate
+
+
+class VsaPredicate(TypedDict):
+    """The 'predicate' field in the Statement layer of a Macaron VSA.
+
+    For reference, see:
+    * in-toto Predicate layer specification:
+    https://github.com/in-toto/attestation/blob/main/spec/v1/predicate.md.
+    * SLSA VSA predicate schema:
+    https://slsa.dev/spec/v1.0/verification_summary#schema.
+    """
+
+    #: Identity of the verifier, as a tool carrying out the verification.
+    verifier: Verifier
+
+    #: The timestamp when the verification occurred.
+    #: The field has the type
+    # `Timestamp <https://github.com/in-toto/attestation/blob/main/spec/v1/field_types.md#Timestamp>`_.
+    timeVerified: str  # noqa: N815
+
+    #: URI that identifies the resource associated with the software component being verified.
+    #: This field has the type
+    #: `ResourceURI <https://github.com/in-toto/attestation/blob/main/spec/v1/field_types.md#ResourceURI>`_.
+    #: Currently, this has the same value as the subject of the VSA, i.e. the PURL of
+    #: the software component being verified against.
+    resourceUri: str  # noqa: N815
+
+    #: The policy that the subject software component was verified against.
+    #: This field has the type
+    #: `ResourceDescriptor <https://github.com/in-toto/attestation/blob/main/spec/v1/resource_descriptor.md>`_.
+    policy: dict[str, Any]
```

Comment:
This attribute needs a more detailed documentation because it is not clear straightaway why we use `policy` as the `ResourceDescriptor`. Also I'm not sure I understand the choice of `dict[str, Any]` type here. Can you please clarify?

Reply:
This field is a `ResourceDescriptor` following [SLSA VSA Predicate schema](https://slsa.dev/spec/v1.0/verification_summary#schema). Currently, it holds only one inner field of `"content"` which holds the Souffle policy code.

I type-annotated it as `dict[str, Any]` since I didn't care to annotate it further, i.e. as long as it is a dictionary with `str` keys, I was fine with it.

I can probably create an additional `Policy(TypedDict)` class with a single `"content"` subfield and annotate the `policy` here with it.

Reply:
See my change in ea5aa606f38abc58d8d91378a6af9e78a7ec8e8b.